crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs

```diff
@@ -235,3 +236,57 @@ console.log(
         config.clone(),
     );
 }
+
+#[test]
+fn test_sourcemap() {
+    let config = ReplaceGlobalDefinesConfig::new(&[
+        ("__OBJECT__", r#"{"hello": "test"}"#),
+        ("__STRING__", r#""development""#),
+        ("__MEMBER__", r"xx.yy.zz"),
+    ])
+    .unwrap();
+    let source_text = r"
+1;
+__OBJECT__;
+2;
+__STRING__;
+3;
+log(__OBJECT__);
+4;
+log(__STRING__);
+5;
+__OBJECT__.hello;
+6;
+log(__MEMBER__);
+7;
+"
+    .trim_start();
+
+    let source_type = SourceType::default();
+    let allocator = Allocator::default();
+    let ret = Parser::new(&allocator, source_text, source_type).parse();
+    let mut program = ret.program;
+    let (symbols, scopes) =
+        SemanticBuilder::new().build(&program).semantic.into_symbol_table_and_scope_tree();
+    let _ = ReplaceGlobalDefines::new(&allocator, config).build(symbols, scopes, &mut program);
+    let result = CodeGenerator::new()
+        .with_options(CodegenOptions {
+            single_quote: true,
+            source_map_path: Some(std::path::Path::new(&"test.js.map").to_path_buf()),
+            ..CodegenOptions::default()
+        })
+        .build(&program);
+
+    let output = result.code;
+    let output_map = result.map.unwrap();
+    let visualizer = SourcemapVisualizer::new(&output, &output_map);
+    let snapshot = visualizer.into_visualizer_text();
+    insta::assert_snapshot!("test_sourcemap", snapshot);
+}
+
+#[test]
+#[should_panic]
+fn test_complex() {
+    let config = ReplaceGlobalDefinesConfig::new(&[("__DEF__", "((() => {})())")]).unwrap();
+    test("__DEF__", "1", config.clone());
+}
```

Comment:
It turned out using more complex expression seems to panic due to lack of scope information. This happens on main too. Here is a stacktrace


```js
---- plugins::replace_global_defines::test_complex stdout ----
thread 'plugins::replace_global_defines::test_complex' panicked at /home/hiroshi/code/others/oxc/crates/oxc_traverse/src/generated/walk.rs:2400:10:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: oxc_traverse::generated::walk::walk_arrow_function_expression
   5: oxc_traverse::generated::walk::walk_expression
   6: oxc_traverse::generated::walk::walk_parenthesized_expression
   7: oxc_traverse::generated::walk::walk_expression
   8: oxc_traverse::generated::walk::walk_call_expression
   9: oxc_traverse::generated::walk::walk_expression
  10: oxc_traverse::generated::walk::walk_parenthesized_expression
  11: oxc_traverse::generated::walk::walk_expression
  12: oxc_traverse::generated::walk::walk_expression_statement
  13: oxc_traverse::generated::walk::walk_statement
  14: oxc_traverse::generated::walk::walk_statements
  15: oxc_traverse::generated::walk::walk_program
  16: oxc_traverse::traverse_mut_with_ctx
  17: oxc_traverse::traverse_mut
  18: oxc_transformer::plugins::replace_global_defines::ReplaceGlobalDefines::build
  19: integrations::plugins::replace_global_defines::test
  20: integrations::plugins::replace_global_defines::test_complex
  21: integrations::plugins::replace_global_defines::test_complex::{{closure}}
  22: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
Oh damn!

That's a separate bug, related to use of `CloneIn`, which removes scope, symbol and reference information from the AST snippet. We don't have a solution to that right now. https://github.com/oxc-project/oxc/issues/4804

I'm going to merge this now and remove the panicking test in a follow-on.

Sorry I wasted your time with that red herring!